src/document_options_save_as_json.lua

```diff
@@ -0,0 +1,709 @@
+    d100 = function(value) return value / 100 end,
+    d10 = function(value) return value / 10 end,
+    d20_48 = function(value) return value / 20.48 end,
+    m100 = function(value) return value * 100 end,
+}
+
+local norm_func_selectors = {
+    FCDistancePrefs = {
+        BarlineDoubleSpace = normalize_funcs.d64,
+        BarlineFinalSpace = normalize_funcs.d64,
+        StemVerticalNoteheadOffset = normalize_funcs.d64
+    },
+    FCGridsGuidesPrefs = {
+        GravityZoneSize = normalize_funcs.d64,
+        GridDistance = normalize_funcs.d64
+    },
+    FCLyricsPrefs = {
+        WordExtLineThickness = normalize_funcs.d64
+    },
+    FCMiscDocPrefs = {
+        FretboardsResizeFraction = normalize_funcs.d10k
+    },
+    FCMusicCharacterPrefs = {
+        DefaultStemLift = normalize_funcs.d64,
+        ["[HV].+Flag[UD]"] = normalize_funcs.d64,        
+    },
+    FCPageFormatPrefs = {
+        SystemStaffHeight = normalize_funcs.d16
+    },
+    FCPianoBracePrefs = {
+        ["."] = normalize_funcs.d10k
+    },
+    FCRepeatPrefs = {
+        ["Thickness$"] = normalize_funcs.d64,
+        SpaceBetweenLines = normalize_funcs.d64,        
+    },
+    FCSizePrefs = {
+        ["Thickness$"] = normalize_funcs.d64,
+        ShapeSlurTipWidth = normalize_funcs.d10k,
+    },
+    FCSmartShapePrefs = {
+        EngraverSlurMaxAngle = normalize_funcs.d100,
+        EngraverSlurMaxLift = normalize_funcs.d64,
+        EngraverSlurMaxStretchFixed = normalize_funcs.d64,
+        EngraverSlurMaxStretchPercent = normalize_funcs.d100,
+        EngraverSlurSymmetryPercent = normalize_funcs.d100,
+        HairpinLineWidth = normalize_funcs.d64,
+        ["^LineWidth$"] = normalize_funcs.d64,
+        SlurTipWidth = normalize_funcs.d10k,
+        Inset = normalize_funcs.d20_48,
+    },
+    FCTiePrefs = {
+        TipWidth = normalize_funcs.d10k,
+        ["tRelativeInset"] = normalize_funcs.m100
+    },
+    FCTupletPrefs = {
+        BracketThickness = normalize_funcs.d64,
+        MaxSlope = normalize_funcs.d10
+    }    
+}
+
+function normalize_units_for_table(t, pattern, func)
+    for key, value in pairs(t) do
+        if type(value) == "table" then 
+            normalize_units_for_table(value, pattern, func)
+        elseif string.find(key, pattern) then
+            t[key] = func(value)
+        end
+    end
+end    
+
+function normalize_units_for_section(section_table, section_tag)
+    local selectors = norm_func_selectors[section_tag]
+    if selectors then
+        for pattern, func in pairs(selectors) do            
+            normalize_units_for_table(section_table, pattern, func)
+        end
+    end
+end
+
+-- endregion
+
+function get_last_key(t)
+    local result
+    for k, _ in pairsbykeys(t) do result = k end
+    return result
+end
+
+-- Simple implementation because we know our inputs are simple
+-- Can't use a json library because we need to keep keys in order
+function get_as_json(t, indent)
```

Comment:
This would be a great function to add to our function library. Not necessary for this PR, but a good future candidate for one.

https://www.finalelua.com/docs/library

Reply:
We'd need to be a little careful. The repo already includes lunajson, which I think is used by one of Robert's scripts, and it can do full serialization and deserialization -- but it doesn't do formatted serialization and can't keep the keys in order (needed in order to be able to diff these files). This little function does formatting and ordering, but it basically assumes that the values are all strings, numbers, bools, or tables. If it's desirable to adopt this into the library, then we should make sure it can handle any values that the library might reasonably throw at it (thinking of arrays, e.g., which are definitely not supported by this function).